### PR TITLE
Fix tuple categorization: add deref adjustments, use index as field name

### LINKED ISSUE #2865: False positive: "Cannot borrow immutable local variable as mutable"
## Environment

* **Intellij-Rust plugin version: 0.2.0.2422-182-nightly**
* **Rust toolchain version: rustc 1.30.0-nightly (cb6d2dfa8 2018-09-16)**
* **IDE name and version: IntelliJ IDEA Ultimate 2018.2.3**
* **Operating system: Ubuntu 16.04 LTS**

## Problem description

Mutably borrowing field from immutably bound mutable struct reference is incorrectly reported as an error.

## Steps to reproduce
```rust
fn main() {
    let mut x = (0, 0);
    let y = &mut x;
    let _z = &mut y.0;
}
```
Cannot borrow immutable local variable `y.0` as mutable

### PR DESCRIPTION
**Pull request by ortem**
Fixes #2865

### CODE REVIEW

File: src/test/kotlin/org/rust/ide/inspections/RsBorrowCheckerInspectionTest.kt [hunk 1]
```diff
@@ -345,6 +345,14 @@ class RsBorrowCheckerInspectionTest : RsInspectionsTestBase(RsBorrowCheckerInspe
         }
     """)
 
+    fun `test tuple`() = checkByText("""
+        fn main() {
+            let mut x = (0, 0);
+            let y = &mut x;
```
Comment (vlad20012) [inline src/test/kotlin/org/rust/ide/inspections/RsBorrowCheckerInspectionTest.kt]: Please, also add a test with multiple derefs.

File: src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt [hunk 2]
```diff
@@ -1181,6 +1181,7 @@ class RsFnInferenceContext(
             for (type in lookup.coercionSequence(receiver)) {
                 if (type is TyTuple) {
                     val fieldIndex = fieldLookup.integerLiteral?.text?.toIntOrNull() ?: return TyUnknown
+                    ctx.addAdjustment(fieldLookup.parentDotExpr.expr, Adjustment.Deref(receiver))
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt]: Multiple derefs may occur here. Is it covered?

**Comment by Undin** [comment]
bors retry

**Pull request merged**